drake/systems/plants/RigidBodyManipulator.cpp

```diff
@@ -2045,26 +2048,35 @@ std::string RigidBodyManipulator::getBodyOrFrameName(int body_or_frame_id)
   if (body_or_frame_id>=0) {
     return bodies[body_or_frame_id]->linkname;
   } else if (body_or_frame_id < -1) {
-    return frames[-body_or_frame_id-2].name;
+    return frames[-body_or_frame_id-2]->name;
   } else {
     return "COM";
   }
 }
 
 template <typename Scalar>
-GradientVar<Scalar, Eigen::Dynamic, 1> RigidBodyManipulator::positionConstraintsNew(int gradient_order)
+GradientVar<Scalar, Eigen::Dynamic, 1> RigidBodyManipulator::positionConstraints(int gradient_order)
 {
   if (gradient_order > 1)
     throw std::runtime_error("only first order gradients are implemented so far (it's trivial to add more)");
 
-  GradientVar<Scalar, Eigen::Dynamic, 1> ret(3*loops.size(), 1, num_positions, gradient_order);
+  GradientVar<Scalar, Eigen::Dynamic, 1> ret(6*loops.size(), 1, num_positions, gradient_order);
   for (size_t i = 0; i < loops.size(); i++) {
-    auto ptA_in_B = forwardKin(loops[i].ptA, loops[i].bodyA->body_index, loops[i].bodyB->body_index, 0, gradient_order);
-
-    ret.value().middleRows(3*i,3) = ptA_in_B.value() - loops[i].ptB;
-
-    if (gradient_order > 0) {
-      ret.gradient().value().middleRows(3*i, 3) = ptA_in_B.gradient().value();
+    { // position constraint
+      auto ptA_in_B = forwardKin(Vector3d::Zero(), loops[i].frameA->frame_index, loops[i].frameB->frame_index, 0,
+                                 gradient_order);
+      ret.value().middleRows(6 * i, 3) = ptA_in_B.value();
```

Comment:
magic number. Also better to use templated (fixed size) version to prevent speed things up and prevent malloc.